File: mcp-server/src/tools/knowledge/get-espionage.ts
/**
 * Espionage tool for AI accessibility (Issue #469)
 */

import { ToolBase } from "../base.js";
import * as z from "zod";
import { ToolAnnotations } from "@modelcontextprotocol/sdk/types.js";
import { getEspionage } from "../../knowledge/getters/espionage.js";
import { MaxMajorCivs } from "../../knowledge/schema/base.js";

const GetEspionageInputSchema = z.object({
  PlayerID: z.number().min(0).max(MaxMajorCivs - 1).optional()
    .describe("Player ID (default: 0)"),
});

const SpySchema = z.object({
  Name: z.string(),
  Rank: z.number(),
  Location: z.string().optional(),
  LocationOwner: z.string().optional(),
  State: z.enum([
    "Unassigned", "Travelling", "GatheringIntel", "CounterIntelligence",
    "Schmoozing", "RiggingElection", "StagingCoup", "Diplomat", "Surveillance", "Unknown"
  ]),
  TurnsInState: z.number().optional(),
});

const CityIntelSchema = z.object({
  CityName: z.string(),
  Owner: z.string(),
  Population: z.number().optional(),
  HasOurSpy: z.boolean(),
  EspionagePotential: z.number().optional().describe("Higher = more valuable target"),
  IsCapital: z.boolean().optional(),
  IsOurCity: z.boolean().optional(),
});

const IntrigueSchema = z.object({
  Turn: z.number(),
  Message: z.string(),
  SpyName: z.string().optional(),
});

const GetEspionageOutputSchema = z.object({
  Spies: z.array(SpySchema),
  CityIntelligence: z.array(CityIntelSchema),
  IntrigueMessages: z.array(IntrigueSchema).describe("Recent intelligence reports"),
});

class GetEspionageTool extends ToolBase {
  readonly name = "get-espionage";

  readonly description = `Get spy locations, states, and intelligence.

Shows:
- All spies with their locations and current activities
- City espionage potential ratings
- Intrigue messages (intelligence reports)
- Counterintelligence status in our cities`;

  readonly inputSchema = GetEspionageInputSchema;
  readonly outputSchema = GetEspionageOutputSchema;

  readonly annotations: ToolAnnotations = { readOnlyHint: true };

  async execute(args: z.infer<typeof this.inputSchema>): Promise<z.infer<typeof this.outputSchema>> {
    const playerID = args.PlayerID ?? 0;
    const data = await getEspionage(playerID);

    if (!data) {
      throw new Error('Failed to get espionage data');
    }

    return data;
  }
}

export default function createGetEspionageTool() {
  return new GetEspionageTool();
}

File: mcp-server/lua/get-espionage.lua
-- Get espionage info for AI accessibility
-- Returns spy locations, states, and intelligence
-- Addresses espionage gap identified in Issue #469

local pPlayer = Players[playerID]
if not pPlayer then
  return nil
end

local result = {
  -- Our spies
  Spies = {},

  -- City intelligence
  CityIntelligence = {},

  -- Intrigue messages
  IntrigueMessages = {},
}

-- Get our spies
local spies = pPlayer:GetEspionageSpies()
if spies then
  for i, spy in ipairs(spies) do
    local spyData = {
      Name = spy.Name,
      Rank = spy.Rank,
    }

    -- Location
    if spy.CityX and spy.CityY then
      local plot = Map.GetPlot(spy.CityX, spy.CityY)
      if plot then
        local city = plot:GetPlotCity()
        if city then
          spyData.Location = city:GetName()
          spyData.LocationOwner = Players[city:GetOwner()] and
                                  Players[city:GetOwner()]:GetCivilizationShortDescription() or "Unknown"
        end
      end
    end

    -- State
    if spy.IsDiplomat then
      spyData.State = "Diplomat"
    elseif spy.EstablishedSurveillance then
      spyData.State = "Surveillance"
    elseif spy.State == 0 then
      spyData.State = "Unassigned"
    elseif spy.State == 1 then
      spyData.State = "Travelling"
    elseif spy.State == 2 then
      spyData.State = "GatheringIntel"
    elseif spy.State == 3 then
      spyData.State = "RiggingElection"
    elseif spy.State == 4 then
      spyData.State = "CounterIntelligence"
    elseif spy.State == 5 then
      spyData.State = "StagingCoup"
    elseif spy.State == 6 then
      spyData.State = "Schmoozing"
    else
      spyData.State = "Unknown"
    end

    -- Progress for certain states
    if spy.TurnsInState then
      spyData.TurnsInState = spy.TurnsInState
    end

    table.insert(result.Spies, spyData)
  end
end

-- Get city espionage potential for cities we know about
for iPlayer = 0, GameDefines.MAX_MAJOR_CIVS - 1 do
  local pLoopPlayer = Players[iPlayer]
  if pLoopPlayer and pLoopPlayer:IsAlive() and iPlayer ~= playerID then
    for city in pLoopPlayer:Cities() do
      -- Check if we've revealed this city
      local plot = city:Plot()
      if plot and plot:IsRevealed(pPlayer:GetTeam(), false) then
        local cityData = {
          CityName = city:GetName(),
          Owner = pLoopPlayer:GetCivilizationShortDescription(),
          Population = city:GetPopulation(),
        }

        -- Check if we have a spy there
        cityData.HasOurSpy = city:HasSpy(playerID)

        -- Get potential (based on population, tech level, etc.)
        -- Higher population = more potential
        cityData.EspionagePotential = city:GetPopulation() * 10

        -- Capital bonus
        if city:IsCapital() then
          cityData.EspionagePotential = cityData.EspionagePotential + 50
          cityData.IsCapital = true
        end

        table.insert(result.CityIntelligence, cityData)
      end
    end
  end
end

-- Get our own cities' counterintelligence
for city in pPlayer:Cities() do
  local cityData = {
    CityName = city:GetName(),
    Owner = "Us",
    HasOurSpy = city:HasSpy(playerID),
    IsOurCity = true,
  }

  table.insert(result.CityIntelligence, cityData)
end

-- Get intrigue messages (last 10)
local intrigueCount = pPlayer:GetNumIntrigueMessages()
local startIndex = math.max(0, intrigueCount - 10)
for i = startIndex, intrigueCount - 1 do
  local intrigue = pPlayer:GetIntrigueMessage(i)
  if intrigue then
    table.insert(result.IntrigueMessages, {
      Turn = intrigue.Turn,
      Message = intrigue.Message,
      SpyName = intrigue.SpyName,
    })
  end
end

return result
